File: Backend/fastApiProject/model/disaster.py
from sqlalchemy import Column, Integer, ForeignKey, Text, Numeric, DateTime, Enum, func
from sqlalchemy.orm import relationship

from config.database import Base
from model.enums import DisasterType, DisasterSeverity, DisasterSource, DisasterStatus


class Disaster(Base):
    __tablename__ = 'disaster'

    id = Column(Integer, primary_key=True)
    created_by = Column(Integer, ForeignKey('users.id'))
    title = Column(Text, nullable=False)
    description = Column(Text, nullable=True)
    type = Column(Enum(DisasterType), nullable=False)
    severity = Column(Enum(DisasterSeverity), nullable=False)
    source = Column(Enum(DisasterSource), nullable=False)
    status = Column(Enum(DisasterStatus), nullable=False)
    radius = Column(Numeric, nullable=False)
    location_id = Column(Integer, ForeignKey('locations.id'))
    external_id = Column(Text, nullable=True, unique=True)
    start_time = Column(DateTime, nullable=False)
    end_time = Column(DateTime, nullable=True)
    created_at = Column(DateTime, nullable=False, server_default=func.now())
    updated_at = Column(DateTime, nullable=False, server_default=func.now(), onupdate=func.now())


    # Relationships
    creator = relationship("User")
    location = relationship("Location")
    assignments = relationship("Assignment", back_populates="disaster")
    resources = relationship("Resource", back_populates="disaster")

File: disaster-live-alerts/src/main/java/com/example/disaster_live_alerts/model/Disaster.java
package com.example.disaster_live_alerts.model;


import com.example.disaster_live_alerts.enums.DisasterSeverity;
import com.example.disaster_live_alerts.enums.DisasterSource;
import com.example.disaster_live_alerts.enums.DisasterStatus;
import com.example.disaster_live_alerts.enums.DisasterType;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Entity
@Table(name = "disaster")
@Data
@AllArgsConstructor
@NoArgsConstructor
@Builder
public class Disaster {
    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE)
    private Long id;

    @JoinColumn(name = "created_by_user_id", nullable = true)
    @ManyToOne(fetch = FetchType.EAGER)
    private User createdBy;

    @Column(nullable = false)
    private String title;

    private String description;

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    private DisasterType type;

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    private DisasterSeverity severity;

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    private DisasterSource source;

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    private DisasterStatus status;

    @Column(nullable = false)
    private Double radius;

    @OneToOne(fetch = FetchType.EAGER)
    private Location location;

    @Column(name = "external_id", nullable = true, unique = true)
    private String externalId;

    @Column(name = "start_time", nullable = false)
    private LocalDateTime startTime;

    @Column(name = "end_time", nullable = true)
    private LocalDateTime endTime;

    @Builder.Default
    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt = LocalDateTime.now();

    @Builder.Default
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt = LocalDateTime.now();
}
